syntax = "proto3";

package backend;

service BackendService {
  // 유저 관리
  rpc GetUser(UserRequest) returns (UserResponse);
  rpc AddUser(NewUser) returns (UserResponse);
  
  // 점수 & 랭킹
  rpc UpdateScore(ScoreUpdate) returns (ScoreResponse);
  rpc GetRanking(RankingRequest) returns (RankingResponse);
}

// 유저 요청 & 응답
message UserRequest {
  string user_id = 1;
}

message NewUser {
  string username = 1;
  string email = 2;
}

message UserResponse {
  string user_id = 1;
  string username = 2;
  string email = 3;
  string created_at = 4;
  bool success = 5;
  string message = 6;
}

// 점수 업데이트
message ScoreUpdate {
  string user_id = 1;
  int32 score = 2;
}

message ScoreResponse {
  string user_id = 1;
  int32 new_score = 2;
  bool success = 3;
  string message = 4;
}

// 랭킹 조회
message RankingRequest {
  int32 limit = 1;
}

message RankingEntry {
  string user_id = 1;
  string username = 2;
  int32 rank = 3;
  int32 score = 4;
}

message RankingResponse {
  repeated RankingEntry rankings = 1;
}